server/src/main/java/org/apache/iotdb/db/qp/strategy/optimizer/ConcatPathOptimizer.java

```diff
@@ -242,6 +244,97 @@ private FilterOperator concatFilter(List<Path> fromPaths, FilterOperator operato
+    // remove stars in fromPaths and get deviceId
+    List<String> noStarDevices = removeStarsInDeviceWithUnique(fromPaths);
+
+    FilterOperator filterTwoFolkTree = new FilterOperator(SQLConstant.KW_OR);
+    FilterOperator currentNode = filterTwoFolkTree;
+    FilterOperator parentNode = currentNode;
+    // to check whether duplicate
+    // e.g. SELECT * FROM root.ln.d1, root.ln.d2 where time < 10
+    // brings two 'time < 10' filter operator
+    Set<FilterOperator> operatorSet = new HashSet<>();
+    for (int i = 0; i < noStarDevices.size(); i++) {
+      FilterOperator newOperator = operator.clone();
+      newOperator = concatFilterPath(noStarDevices.get(i), newOperator);
+
+      if (!operatorSet.contains(newOperator)) {
+        if(currentNode.getChildren().size() > 0){
+          FilterOperator newInnerNode = new FilterOperator(SQLConstant.KW_OR);
+          currentNode.addChildOperator(newInnerNode);
+          parentNode = currentNode;
+          currentNode = newInnerNode;
+        }
+        currentNode.addChildOperator(newOperator);
+        operatorSet.add(newOperator);
+      }
+    }
+
+    // if 'OR' has no enough operands due to duplication
+    if(currentNode.getChildren().size() == 1){
+      if(parentNode == currentNode){
+        filterTwoFolkTree = currentNode.getChildren().get(0);
+      } else {
+        parentNode.getChildren().set(1, currentNode.getChildren().get(0));
+      }
+    }
+
+    return filterTwoFolkTree;
+  }
+
+  private List<String> removeStarsInDeviceWithUnique(List<Path> paths)
+          throws LogicalOptimizeException {
+    List<String> retDevices;
+    HashSet<String> deviceSet = new HashSet<>();
+    try {
+      for (Path path : paths) {
+        List<String> all;
+        all = executor.getAllPaths(path.getFullPath());
+        if (all.isEmpty()) {
+          throw new LogicalOptimizeException(
+                  "Path: \"" + path + "\" doesn't correspond to any known time series");
+        }
+        for (String subPath : all) {
+          String deviceId = new Path(subPath).getDevice();
+          if (!deviceSet.contains(deviceId)) {
+            deviceSet.add(deviceId);
+          }
+        }
+      }
+
+      retDevices = new ArrayList<>(deviceSet);
+    } catch (MetadataException e) {
+      throw new LogicalOptimizeException("error when remove star: " + e.getMessage());
+    }
+    return retDevices;
+  }
+
+  private FilterOperator concatFilterPath(String prefix, FilterOperator operator) {
+    if(!operator.isLeaf()){
+      for (FilterOperator child : operator.getChildren()) {
+        concatFilterPath(prefix, child);
+      }
+      return operator;
+    }
+    BasicFunctionOperator basicOperator = (BasicFunctionOperator) operator;
+    Path filterPath = basicOperator.getSinglePath();
+
+    // do nothing in the cases of "where time > 5" or "where root.d1.s1 > 5"
+    if (SQLConstant.isReservedPath(filterPath) || filterPath.startWith("ROOT")) {
```

Comment:
So all paths in the filter have been cast to upper case? It is new knowledge to me, I just want to confirm.

Reply:
It seems not. I used the existing code before. I will fix it all.